### DISCUSSION TITLE
Support for raspberypi

### DISCUSSION THREAD

**Discussion by marimeireles** [pyscript/pyscript#1935] [Proposals]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a duplication
- [X] I already searched in Google and didn't find any good information or help

### What is the issue/comment/problem?

Folks are interested in ways we can interact with a Rasppi!
A reason to deepen our relationship with the embedded ecosystem and specially with the RaspberryPi foundation is that one of their main goals is very aligned to ours: we both have deep interest in pursuing educational development through the empowerment of people who are not the every day users of programming languages.

What are our current short term goals and what we hope to achieve?

We want to control a raspi from the browser with PyScript. We decided to give it a go using [gpiozero](https://gpiozero.readthedocs.io/en/stable/index.html) and luckily @ntoll connected us with one of the maintainers for a call!

Unfortunately, it seems like it's not possible to simply use a REST API to directly communicate with the rasppi because of the way the Raspberry TCP works. @antocuni read more into it and could probably offer better clarification to the issue.

After discussing with Ben Nuttal the team decided to use [pgpio](https://github.com/joan2937/pigpio) as an approach to communicate with the Rasppi from the browser. (On a side note, Ben told us the maintainer of pgpio, https://github.com/joan2937/ is a very responsive person on IRC and might be willing to help us here!)
The reason for using pgpio being, according to Ben, we might face really complex use cases that we're not anticipating when trying to communicate with embedded devices.

I think as a first try to this issue we need to:
- [ ] have a working server running either on the raspi or the desktop
- [ ] API that will simply POST and GET messages (here we should probably use Flask?)
- [ ] have a protocol communication layer that will be serialized in an legible format for pyscript

Some questions I still have are:
a) is not clear to me if we'll still need websockify if we're using pgpio (I guess that's the point of using it, right?)
b) how does the interaction with gpiozero and pgpio works, is pgpio fully supported?

Some ideas for the future:
- implement a functional raspi on the browser, for that we'd use the [mockAPI](https://gpiozero.readthedocs.io/en/stable/api_pins.html#mock-pins)

**Comment by WebReflection** [comment]
FWIW, I literally *could* do this via PyScript and a [coincident/server like](https://github.com/WebReflection/coincident?tab=readme-ov-file#coincidentserver) module.

To make it happen though, I need (out of a quick thoughts):

  * a Pyodide or MicroPython module that doesn't try to reach the device but uses *coincident* behind the scene ... on the browser side
  * a Pyodide or MicroPython or pure Python that, on the server side, understands WebSocket and handles properly requests from the browser (including cross socket GC primitives)

I have this working already in a JS to JS world, with NodeJS, Deno or Bun being the "actors" on the server side, so I don't see why this couldn't be ported or done in a Python world too.

I can see a lot of potentials but also a lot of time behind to have the whole thing working, yet I think this might be just awesome!

I think this also should be moved into discussions as it's easier to follow up comments and branch out.

**Reply by WebReflection** [reply]
if interested about the current state: https://twitter.com/WebReflection/status/1678762388538155013
